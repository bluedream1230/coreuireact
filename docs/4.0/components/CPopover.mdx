---
title: React Popover Component
name: Popover
description: Documentation and examples for adding React popovers, like those found in iOS, to any element on your site.

menu: Components
route: /components/popover
---

import { Playground, Props } from 'docz'
import { CPopover, CButton } from '../../../src/index.ts'

## Example

<Playground>
  <CPopover
    title="Popover title"
    content="And here’s some amazing content. It’s very engaging. Right?"
    placement="right"
  >
    <CButton color="danger" size="lg">Click to toggle popover</CButton>
  </CPopover>
</Playground>

### Four directions

Four options are available: top, right, bottom, and left aligned. Directions are mirrored when using CoreUI for React in RTL.

<Playground>
  <CPopover
    content="Vivamus sagittis lacus vel augue laoreet rutrum faucibus."
    placement="top"
  >
    <CButton color="secondary">Popover on top</CButton>
  </CPopover>
  <CPopover
    content="Vivamus sagittis lacus vel augue laoreet rutrum faucibus."
    placement="right"
  >
    <CButton color="secondary">Popover on right</CButton>
  </CPopover>
  <CPopover
    content="Vivamus sagittis lacus vel augue laoreet rutrum faucibus."
    placement="bottom"
  >
    <CButton color="secondary">Popover on bottom</CButton>
  </CPopover>
  <CPopover
    content="Vivamus sagittis lacus vel augue laoreet rutrum faucibus."
    placement="left"
  >
    <CButton color="secondary">Popover on left</CButton>
  </CPopover>
</Playground>

## API

### CPopover

<Props of={CPopover} />