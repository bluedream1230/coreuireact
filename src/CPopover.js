import React, {useState} from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import CTooltipPopoverWrapper from './CTooltipPopoverWrapper';

//component - CoreUI / CPopover

const CPopover = props=>{

  let {
    custom,
    //
    toggle,
    show,
    defaultOpen,
    ...attributes
  } = props;

  const [isOpen, setIsOpen] = useState(defaultOpen || false);

  if (!custom){
    const userToggle = toggle;
    toggle = (e)=>{
      setIsOpen(!isOpen);
      if (userToggle)
        userToggle(e);
    }
    show = isOpen;
  }

  //render

  const popperClasses = classNames(
    'popover',
    'show',
    props.className
  );

  const classes = classNames(
    'popover-inner',
    props.innerClassName
  );

  return (
    <CTooltipPopoverWrapper
      {...attributes}
      toggle={toggle}
      isOpen={show}
      className={popperClasses}
      innerClassName={classes}
    />
  );

}

CPopover.propTypes = {
  ...CTooltipPopoverWrapper.propTypes,
  custom: PropTypes.bool,
  //
  innerRef: PropTypes.oneOfType([PropTypes.object, PropTypes.func, PropTypes.string]),
  defaultOpen: PropTypes.bool,
}

CPopover.defaultProps = {
  placement: 'right',
  placementPrefix: 'bs-popover',
  trigger: 'click',
};

export default CPopover;
