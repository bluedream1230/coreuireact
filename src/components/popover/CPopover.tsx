// TODO: fix possition

import React, { FC, ReactNode, useState } from 'react'
import { createPortal } from 'react-dom'
import PropTypes from 'prop-types'
import { Manager, Popper, Reference } from 'react-popper'
import { CSSTransition } from 'react-transition-group'

import { CPopoverContent } from './CPopoverContent'
import { Placements, Triggers, triggerPropType } from '../Types'

export interface CPopoverProps {
  children: JSX.Element
  /**
   * Content node for your component. [docs]
   */
  content: ReactNode
  /**
   * Title node for your component. [docs]
   */
  title?: ReactNode
  /**
   * Sets which event handlers youâ€™d like provided to your toggle prop. You can specify one trigger or an array of them. [docs]
   */
  trigger?: Triggers | Triggers[]
  /**
   * Describes the placement of your component after Popper.js has applied all the modifiers that may have flipped or altered the originally provided placement property. [docs]
   *
   * @type 'auto' | 'top-end' | 'top' | 'top-start' | 'bottom-end' | 'bottom' | 'bottom-start' | 'right-start' | 'right' | 'right-end' | 'left-start' | 'left' | 'left-end'
   * @default 'bottom'
   */
  placement?: Placements | 'start' | 'end' // TODO: find solution
  /**
   * Toggle the visibility of popover component. [docs]
   *
   * @default true
   */
  visible?: boolean
}

export const CPopover: FC<CPopoverProps> = ({
  children,
  placement = 'top',
  trigger = 'click',
  visible,
  ...rest
}) => {
  const [_visible, setVisible] = useState(visible)
  const _placement = placement === 'start' ? 'left' : placement === 'end' ? 'right' : placement

  const getTransitionClass = (state: string) => {
    return state === 'entering'
      ? 'fade'
      : state === 'entered'
      ? 'fade show'
      : state === 'exiting'
      ? 'fade'
      : 'fade'
  }

  return (
    <Manager>
      <Reference>
        {({ ref }) =>
          React.cloneElement(children, {
            ref: ref,
            ...((trigger === 'click' || trigger.includes('click')) && {
              onClick: () => setVisible(!_visible),
            }),
            ...((trigger === 'focus' || trigger.includes('focus')) && {
              onFocus: () => setVisible(true),
              onBlur: () => setVisible(false),
            }),
            ...((trigger === 'hover' || trigger.includes('hover')) && {
              onMouseEnter: () => setVisible(true),
              onMouseLeave: () => setVisible(false),
            }),
          })
        }
      </Reference>
      {typeof window !== 'undefined' &&
        createPortal(
          <CSSTransition
            in={_visible}
            timeout={{
              enter: 0,
              exit: 200,
            }}
            mountOnEnter
            unmountOnExit
          >
            {(state) => {
              const transitionClass = getTransitionClass(state)
              return (
                <Popper placement={_placement}>
                  {(p) => (
                    <CPopoverContent
                      transitionClass={transitionClass}
                      placementClassNamePostfix={placement}
                      {...rest}
                      {...p}
                    ></CPopoverContent>
                  )}
                </Popper>
              )
            }}
          </CSSTransition>,
          document.body,
        )}
    </Manager>
  )
}

CPopover.propTypes = {
  children: PropTypes.any,
  placement: PropTypes.oneOf([
    'auto',
    'top-end',
    'top',
    'top-start',
    'bottom-end',
    'bottom',
    'bottom-start',
    'right-start',
    'right',
    'right-end',
    'left-start',
    'left',
    'left-end',
    'start',
    'end',
  ]),
  trigger: triggerPropType,
  visible: PropTypes.bool,
}

CPopover.displayName = 'CPopover'
