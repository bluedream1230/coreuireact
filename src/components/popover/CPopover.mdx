---
title: React Popover Component
name: Popover
description: Documentation and examples for adding React popovers, like those found in iOS, to any element on your site.

menu: Components
route: /components/popover
---

import { Playground, Props } from 'docz'
import { CPopover } from './CPopover.tsx'
import { CButton } from './../button/CButton.tsx'

## Example

<Playground>
  <CPopover
    title="Popover title"
    content="And here’s some amazing content. It’s very engaging. Right?"
    placement="end"
  >
    <CButton color="danger" size="lg">Click to toggle popover</CButton>
  </CPopover>
</Playground>

### Four directions

Four options are available: top, right, bottom, and left aligned. Directions are mirrored when using CoreUI for React in RTL.

<Playground>
  <CPopover
    content="Vivamus sagittis lacus vel augue laoreet rutrum faucibus."
    placement="top"
  >
    <CButton color="secondary">Popover on top</CButton>
  </CPopover>
  <CPopover
    content="Vivamus sagittis lacus vel augue laoreet rutrum faucibus."
    placement="end"
  >
    <CButton color="secondary">Popover on right</CButton>
  </CPopover>
  <CPopover
    content="Vivamus sagittis lacus vel augue laoreet rutrum faucibus."
    placement="bottom"
  >
    <CButton color="secondary">Popover on bottom</CButton>
  </CPopover>
  <CPopover
    content="Vivamus sagittis lacus vel augue laoreet rutrum faucibus."
    placement="start"
  >
    <CButton color="secondary">Popover on left</CButton>
  </CPopover>
</Playground>

## API

### CPopover

<Props of={CPopover} />