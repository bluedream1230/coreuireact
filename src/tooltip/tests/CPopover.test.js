import React from 'react'
import { configure, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import renderer from 'react-test-renderer';

import CPopover from '../CPopover'

configure({ adapter: new Adapter() })

describe('CPopover', () => {
  it('renders customized wrapper correctly', () => {
    const componentCustomized = renderer.create(
      <CPopover
        header='header'
        content='content'
      >
        <a href="#"> Popover example </a>
      </CPopover>
    );
    let tree = componentCustomized.toJSON();
    expect(tree).toMatchSnapshot()
  })
})