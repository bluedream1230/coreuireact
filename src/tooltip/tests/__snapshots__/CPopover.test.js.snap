// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CPopover renders customized wrapper correctly 1`] = `
<a
  data-tooltip="2jynnlorl27"
  href="#"
>
   Popover example 
</a>
`;
